File: ios/Classes/Messages.swift
//
//  Messages.swift
//  flutter_qcloudcosxml
//
//  Created by Haijun on 2021/4/30.
//

import Flutter
import Foundation

struct QCloudCosInitOptions {
    var region: String
    var secretId: String?
    var secretKey: String?
    var appId: String?
}

extension QCloudCosInitOptions {
    init(map: [String: Any]) {
        self.init(
            region: map["region"] as? String ?? "",
            secretId: map["secretId"] as? String,
            secretKey: map["secretKey"] as? String,
            appId: map["appId"] as? String
        )
    }

    func toMap() -> [String: Any] {
        return [
            "region": region,
            "secretId": secretId ?? NSNull(),
            "secretKey": secretKey ?? NSNull(),
            "appId": appId ?? NSNull(),
        ]
    }
}

struct QCloudCosManagerUploadOptions {
    var filePath: String
    var savePath: String
    var bucket: String
}

extension QCloudCosManagerUploadOptions {
    init(map: [String: Any]) {
        self.init(
            filePath: map["filePath"] as? String ?? "",
            savePath: map["savePath"] as? String ?? "",
            bucket: map["bucket"] as? String ?? ""
        )
    }

    func toMap() -> [String: Any] {
        return [
            "filePath": filePath,
            "savePath": savePath,
            "bucket": bucket,
        ]
    }
}

struct QCloudCosManagerUploadResult {
    var key: String
    var location: String
}

extension QCloudCosManagerUploadResult {
    init(map: [String: Any]) {
        self.init(
            key: map["key"] as? String ?? "",
            location: map["location"] as? String ?? ""
        )
    }

    func toMap() -> [String: Any] {
        return [
            "key": key,
            "location": location,
        ]
    }
}

protocol QCloudCosManager {
    func initialize(input: QCloudCosInitOptions, completion: @escaping (_ error: FlutterError?) -> Void)
    func upload(input: QCloudCosManagerUploadOptions, completion: @escaping (_ result: QCloudCosManagerUploadResult?, _ error: FlutterError?) -> Void)
}

func FlutterQCloudCOSXMLSetup(binaryMessenger: FlutterBinaryMessenger, api: QCloudCosManager) {
    let channel = FlutterBasicMessageChannel(name: "dev.flutter.pigeon.QCloudCosManager.initialize", binaryMessenger: binaryMessenger)
    channel.setMessageHandler { message, callback in
        let input = QCloudCosInitOptions(map: message as! [String : Any])
        api.initialize(input: input) { error in
            callback(wrapResult(result: nil, error: error))
        }
    }
    
    let channel1 = FlutterBasicMessageChannel(name: "dev.flutter.pigeon.QCloudCosManager.upload", binaryMessenger: binaryMessenger)
    channel1.setMessageHandler { message, callback in
        let input = QCloudCosManagerUploadOptions(map: message as! [String : Any])
        api.upload(input: input) { result, error in
            callback(wrapResult(result: result?.toMap(), error: error))
        }
    }
}

File: android/src/main/kotlin/com/haijunwei/flutter_qcloudcosxml/Messages.kt
package com.haijunwei.flutter_qcloudcosxml

import com.haijunwei.flutter_qcloudcosxml.bean.QCloudCosInitOptions
import com.haijunwei.flutter_qcloudcosxml.bean.QCloudCosManagerUploadOptions
import com.haijunwei.flutter_qcloudcosxml.bean.QCloudCosManagerUploadResult
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import java.util.*

/** Generated class from Pigeon.  */
object Messages {
    /** Sets up an instance of `PhotoPicker` to handle messages through the `binaryMessenger`.  */
    fun setup(binaryMessenger: BinaryMessenger?, api: QCloudCosXml?) {
        run {
            initialize(binaryMessenger,api)
            upload(binaryMessenger,api)
        }
    }

    private fun initialize(binaryMessenger: BinaryMessenger?, api: QCloudCosXml?){
        val channel = BasicMessageChannel(binaryMessenger!!, "dev.flutter.pigeon.QCloudCosManager.initialize", StandardMessageCodec())
        if (api == null) {
            channel.setMessageHandler(null)
            return
        }
        channel.setMessageHandler { message: Any?, reply: BasicMessageChannel.Reply<Any> ->
            val wrapped: MutableMap<String, Any> = HashMap()
            try {
                val input = QCloudCosInitOptions.fromMap(message as Map<String?, Any?>?)
                api.initialize(input, object : Result<Any?> {
                    override fun success(result: Any?) {
                        val map:MutableMap<String, Any?> = HashMap()
                        map["message"] = "success"
                        wrapped["result"] = map
                        reply.reply(wrapped)
                    }

                    override fun error(throwable: Throwable) {
                        wrapped["error"] = wrapError(throwable)
                        reply.reply(wrapped)
                    }
                })
            } catch (exception: Error) {
                wrapped["error"] = wrapError(exception)
                reply.reply(wrapped)
            } catch (exception: RuntimeException) {
                wrapped["error"] = wrapError(exception)
                reply.reply(wrapped)
            }
        }
    }

    private fun upload(binaryMessenger: BinaryMessenger?, api: QCloudCosXml?){
        val channel = BasicMessageChannel(binaryMessenger!!, "dev.flutter.pigeon.QCloudCosManager.upload", StandardMessageCodec())
        if (api == null) {
            channel.setMessageHandler(null)
            return
        }

        channel.setMessageHandler { message: Any?, reply: BasicMessageChannel.Reply<Any> ->
            val wrapped: MutableMap<String, Any> = HashMap()
            try {
                val input = QCloudCosManagerUploadOptions.fromMap(message as Map<String?, Any?>?)
                api.upload(input, object : Result<QCloudCosManagerUploadResult?> {
                    override fun success(result: QCloudCosManagerUploadResult?) {
                        wrapped["result"] = result?.toMap()?:""
                        reply.reply(wrapped)
                    }

                    override fun error(throwable: Throwable) {
                        wrapped["error"] = wrapError(throwable)
                        reply.reply(wrapped)
                    }
                })
            } catch (exception: Error) {
                wrapped["error"] = wrapError(exception)
                reply.reply(wrapped)
            } catch (exception: RuntimeException) {
                wrapped["error"] = wrapError(exception)
                reply.reply(wrapped)
            }
        }
    }
}

private fun wrapError(exception: Throwable): Map<String, Any?> {
    val errorMap: MutableMap<String, Any?> = HashMap()
    errorMap["message"] = exception.toString()
    errorMap["code"] = exception.javaClass.simpleName
    errorMap["details"] = null
    return errorMap
}


interface Result<T> {
    fun success(result: T)

    fun error(throwable:Throwable)
}

/** Generated interface from Pigeon that represents a handler of messages from Flutter. */
interface QCloudCosXml {
    fun upload(options: QCloudCosManagerUploadOptions?, result: Result<QCloudCosManagerUploadResult?>?)
    fun initialize(options: QCloudCosInitOptions?, result: Result<Any?>?)
}